libraries/botframework-config/src/botConfiguration.ts

```diff
@@ -0,0 +1,230 @@
+                    .toArray();
+            }
+        }
+
+        if (hasSecret)
+            this.encrypt(secret);
+
+        await fsx.writeJson(botpath || <string>this.internal.location, this.toJSON(), { spaces: 4 });
+
+        if (hasSecret)
+            this.decrypt(secret);
+    }
+
+    public clearSecret() {
+        this.secretKey = '';
+    }
+
+    // connect to a service
+    public connectService(newService: IConnectedService): void {
+        if (Enumerable.fromSource(this.services)
+            .where(s => s.type == newService.type)
+            .where(s => s.id == newService.id)
+            .any()) {
+            throw Error(`service with ${newService.id} already connected`);
+        } else {
+            // give unique name
+            let nameCount = 1;
+            let name = newService.name;
+
+            while (true) {
+                if (nameCount > 1) {
+                    name = `${newService.name} (${nameCount})`;
+                }
+
+                if (!Enumerable.fromSource(this.services).where(s => s.name == name).any())
+                    break;
+                nameCount++;
+            }
+            newService.name = name;
+
+            this.services.push(BotConfiguration.serviceFromJSON(newService));
+        }
+    }
+
+    // encrypt all values in the config
+    public encrypt(secret: string) {
+        this.validateSecretKey(secret);
+
+        for (let service of this.services) {
+            (<ConnectedService>service).encrypt(secret);
+        }
+    }
+
+    // decrypt all values in the config
+    public decrypt(secret?: string) {
+        this.validateSecretKey(secret);
+
+        for (let service of this.services) {
+            (<ConnectedService>service).decrypt(secret);
+        }
+    }
+
+
+    // remove service by name or id
+    public disconnectServiceByNameOrId(nameOrId: string): IConnectedService {
+        let svs = new List<IConnectedService>(this.services);
+
+        for (let i = 0; i < svs.count(); i++) {
+            let service = svs.elementAt(i);
+            if (service.id == nameOrId || service.name == nameOrId) {
+                svs.removeAt(i);
+                this.services = svs.toArray();
+                return service;
+            }
+        }
+        throw new Error(`a service with id or name of [${nameOrId}] was not found`);
+    }
+
+    // remove a service
+    public disconnectService(type: string, id: string): void {
+        let svs = new List<IConnectedService>(this.services);
+
+        for (let i = 0; i < svs.count(); i++) {
+            let service = svs.elementAt(i);
+            if (service.type == type && service.id == id) {
+                svs.removeAt(i);
+                this.services = svs.toArray();
+                return;
+            }
+        }
+    }
+
+
+    // make sure secret is correct by decrypting the secretKey with it
+    public validateSecretKey(secret: string): void {
+        if (!secret || secret.length == 0) {
```

Comment:
a string with a length of zero resolves to false. It's safe to remove `|| string.length == 0`